// @esri/calcite-components/dist/components/calcite-tip-group/customElement.js
/*! All material copyright ESRI, All Rights Reserved, unless otherwise specified.
See https://github.com/Esri/calcite-design-system/blob/dev/LICENSE.md for details.
v3.3.3 */
import { c as customElement } from "../../chunks/runtime.js";
import { css, html } from "lit";
import { LitElement } from "@arcgis/lumina";
import { l as logger } from "../../chunks/logger.js";
const styles = css`:host{--calcite-internal-color-focus: var( --calcite-color-focus, var(--calcite-ui-focus-color, var(--calcite-color-brand)) )}:host{box-sizing:border-box;display:block;background-color:var(--calcite-color-foreground-1);font-size:var(--calcite-font-size--1);line-height:1rem;color:var(--calcite-color-text-2)}::slotted(calcite-tip){margin:0;border-style:none;max-inline-size:var(--calcite-tip-max-width)}:host([hidden]){display:none}[hidden]{display:none}`;
class TipGroup extends LitElement {
  static {
    this.properties = { groupTitle: 1 };
  }
  static {
    this.styles = styles;
  }
  load() {
    logger.deprecated("component", {
      name: "tip-group",
      removalVersion: 4,
      suggested: ["carousel", "carousel-item"]
    });
  }
  render() {
    return html`<slot></slot>`;
  }
}
customElement("calcite-tip-group", TipGroup);
export {
  TipGroup
};
