// @esri/calcite-components/dist/chunks/resources6.js
/*! All material copyright ESRI, All Rights Reserved, unless otherwise specified.
See https://github.com/Esri/calcite-design-system/blob/dev/LICENSE.md for details.
v3.3.3 */
const ComboboxItemSelector = "CALCITE-COMBOBOX-ITEM";
const ComboboxItemGroupSelector = "CALCITE-COMBOBOX-ITEM-GROUP";
const AllComboboxChildrenSelector = `${ComboboxItemSelector}, ${ComboboxItemGroupSelector}`;
const CSS = {
  allSelected: "all-selected",
  chip: "chip",
  chipInvisible: "chip--invisible",
  icon: "icon",
  input: "input",
  inputHidden: "input--hidden",
  label: "label",
  labelIcon: "label--icon",
  listContainer: "list-container",
  noMatches: "no-matches",
  noMatchesPlaceholder: "no-matches-placeholder",
  placeholderIcon: "placeholder-icon",
  selectAll: "select-all",
  selectionDisplayFit: "selection-display--fit",
  selectionDisplaySingle: "selection-display--single",
  selectedIcon: "selected-icon",
  floatingUIContainer: "floating-ui-container",
  screenReadersOnly: "screen-readers-only",
  wrapper: "wrapper",
  wrapperSingle: "wrapper--single",
  wrapperActive: "wrapper--active",
  gridInput: "grid-input",
  inputSingle: "input--single",
  inputIcon: "input--icon",
  inputWrap: "input-wrap",
  inputWrapSingle: "input-wrap--single",
  iconEnd: "icon-end",
  iconStart: "icon-start",
  list: "list",
  listHide: "list--hide"
};
const idPrefix = "combobox";
const IDS = {
  validationMessage: "comboboxValidationMessage",
  item: (id) => `${idPrefix}-item-${id}`,
  chip: (id) => `${idPrefix}-chip-${id}`,
  label: (id) => `${idPrefix}-label-${id}`,
  listbox: (id) => `${idPrefix}-listbox-${id}`,
  input: (id) => `${idPrefix}-input-${id}`
};
const ICONS = {
  chevronUp: "chevron-up",
  chevronDown: "chevron-down"
};
const SLOTS = {
  labelContent: "label-content"
};
export {
  AllComboboxChildrenSelector as A,
  ComboboxItemSelector as C,
  IDS as I,
  SLOTS as S,
  ComboboxItemGroupSelector as a,
  CSS as b,
  ICONS as c
};
