// @esri/calcite-components/dist/chunks/resources3.js
/*! All material copyright ESRI, All Rights Reserved, unless otherwise specified.
See https://github.com/Esri/calcite-design-system/blob/dev/LICENSE.md for details.
v3.3.3 */
const CSS = {
  actionBarContainer: "action-bar-container",
  container: "container",
  contentBottom: "content-bottom",
  contentTop: "content-top",
  header: "header",
  headerContainer: "header-container",
  headerContainerBorderEnd: "header-container--border-end",
  headingTextContent: "heading-text-content",
  heading: "heading",
  description: "description",
  headerContent: "header-content",
  headerActions: "header-actions",
  headerActionsEnd: "header-actions--end",
  headerActionsStart: "header-actions--start",
  icon: "icon",
  contentWrapper: "content-wrapper",
  fabContainer: "fab-container",
  footer: "footer",
  footerContent: "footer-content",
  footerActions: "footer-actions",
  footerStart: "footer-start",
  footerEnd: "footer-end",
  headerSlottedContent: "header--slotted-content",
  headerNonSlottedContent: "header--non-slotted-content",
  menuAction: "menu-action"
};
const IDS = {
  close: "close",
  collapse: "collapse"
};
const ICONS = {
  close: "x",
  menu: "ellipsis",
  expand: "chevron-down",
  collapse: "chevron-up"
};
const SLOTS = {
  actionBar: "action-bar",
  alerts: "alerts",
  contentBottom: "content-bottom",
  contentTop: "content-top",
  headerActionsStart: "header-actions-start",
  headerActionsEnd: "header-actions-end",
  headerMenuActions: "header-menu-actions",
  headerContent: "header-content",
  fab: "fab",
  footer: "footer",
  footerEnd: "footer-end",
  footerStart: "footer-start",
  footerActions: "footer-actions"
};
export {
  CSS as C,
  ICONS as I,
  SLOTS as S,
  IDS as a
};
