// @esri/calcite-components/dist/components/calcite-tile-select-group/customElement.js
/*! All material copyright ESRI, All Rights Reserved, unless otherwise specified.
See https://github.com/Esri/calcite-design-system/blob/dev/LICENSE.md for details.
v3.3.3 */
import { c as customElement } from "../../chunks/runtime.js";
import { css, html } from "lit";
import { LitElement } from "@arcgis/lumina";
import { u as updateHostInteraction, I as InteractiveContainer } from "../../chunks/interactive.js";
import { l as logger } from "../../chunks/logger.js";
const styles = css`:host{--calcite-internal-color-focus: var( --calcite-color-focus, var(--calcite-ui-focus-color, var(--calcite-color-brand)) )}:host([disabled]){cursor:default;-webkit-user-select:none;user-select:none;opacity:var(--calcite-opacity-disabled)}:host([disabled]) *,:host([disabled]) ::slotted(*){pointer-events:none}:host{display:flex;flex-wrap:wrap}:host ::slotted(calcite-tile-select){margin-block-end:1px;margin-inline-end:1px}:host([layout=vertical]){flex-direction:column}:host([disabled]) ::slotted([calcite-hydrated][disabled]),:host([disabled]) [calcite-hydrated][disabled]{opacity:1}.interaction-container{display:contents}:host([hidden]){display:none}[hidden]{display:none}`;
class TileSelectGroup extends LitElement {
  constructor() {
    super(...arguments);
    this.disabled = false;
    this.layout = "horizontal";
  }
  static {
    this.properties = { disabled: [7, {}, { reflect: true, type: Boolean }], layout: [3, {}, { reflect: true }] };
  }
  static {
    this.styles = styles;
  }
  load() {
    logger.deprecated("component", {
      name: "tile-select-group",
      removalVersion: 4,
      suggested: ["tile", "tile-group"]
    });
  }
  updated() {
    updateHostInteraction(this);
  }
  render() {
    return InteractiveContainer({ disabled: this.disabled, children: html`<slot></slot>` });
  }
}
customElement("calcite-tile-select-group", TileSelectGroup);
export {
  TileSelectGroup
};
