// @esri/calcite-components/dist/components/calcite-option-group/customElement.js
/*! All material copyright ESRI, All Rights Reserved, unless otherwise specified.
See https://github.com/Esri/calcite-design-system/blob/dev/LICENSE.md for details.
v3.3.3 */
import { c as customElement } from "../../chunks/runtime.js";
import { css, html } from "lit";
import { LitElement, createEvent } from "@arcgis/lumina";
const styles = css`:host{--calcite-internal-color-focus: var( --calcite-color-focus, var(--calcite-ui-focus-color, var(--calcite-color-brand)) )}:host{display:block}:host([hidden]){display:none}[hidden]{display:none}`;
class OptionGroup extends LitElement {
  constructor() {
    super(...arguments);
    this.disabled = false;
    this.calciteInternalOptionGroupChange = createEvent({ cancelable: false });
  }
  static {
    this.properties = { disabled: [7, {}, { reflect: true, type: Boolean }], label: 1 };
  }
  static {
    this.styles = styles;
  }
  willUpdate(changes) {
    if (changes.has("disabled") && (this.hasUpdated || this.disabled !== false) || changes.has("label")) {
      this.calciteInternalOptionGroupChange.emit();
    }
  }
  render() {
    return html`<div>${this.label}</div><slot></slot>`;
  }
}
customElement("calcite-option-group", OptionGroup);
export {
  OptionGroup
};
